### PR TITLE
add notes for n17526

### PR DESCRIPTION
**Pull request by glozow**

### CODE REVIEW

**Comment by jonatack** [comment]
(The file needs to be in the `_posts` dir...is the rake task not working?)

```rake
rake posts:new -- --pr NUMBER --host USERNAME --date YYYY-MM-DD
```

File: 2021-05-05-#17526.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	  `CoinEligibilityFilter`s are used to filter these coins within `SelectCoinsMinConf()`.
+ We have a clear hierarchy of which Coin Selection solutions are preferred: we first try with a
+restriction of at least 6 confirmations on foreign UTXOs and 1 confirmation on our own UTXOs.
+If no solution is found, we try with at least 1 confirmation on all UTXOs. If that doesn't work and
+the user allows spending unconfirmed change, we try that (still requiring at least 1 confirmation on
+foreign UTXOs), gradually increasing mempool ancestor limits on the unconfirmed change.
+
+	- Within `SelectCoinsMinConf()`, `OutputGroup`s are created using the
+	  `CoinEligibilityFilter`. There is a clear preference for BnB and we will only try Knapsack
+and SRD if that fails. However, we'll try both Knapsack and SRD together, picking the solution with
```
Comment (Xekyo) [inline 2021-05-05-#17526.md:44]: Sentence misses end.

File: 2021-05-05-#17526.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+3. Why might we prefer to spend more inputs in the same transaction?
+
+4. Quiz: Based on the Coin Selection scheme proposed here, let's say that Solutions A, B, and C
+   exist (ignore the fact that we would exit early after finding a solution we're satisfied with). Which would we pick?
+
+	 Solution A: picked using Knapsack. Produces a change output, pays 100 satoshis in fees, and
+only uses confirmed UTXOs, each with 4 confirmations.
+
+	Solution B: picked using BnB. No change output, pays 95 satoshis in fees, and uses one
```
Comment (Xekyo) [inline 2021-05-05-#17526.md:62]: Mh, I dimly remember that BnB never uses unconfirmed inputs, but maybe that was a different codebase. ;)

File: 2021-05-05-#17526.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+confirmed UTXOs, each with 1 confirmation.
+
+
+5. What are `OutputGroups`? Why does SRD pick from output groups rather than from UTXOs?
+
+6. What does calling `GroupOutputs` with `positive_only=true` do (Hint: you may
+   want to review what effective values are)? What could happen if
+`SelectCoinsSRD()` was called with `all_groups` instead of `positive_groups`?
+
+7. What are some ways a deterministic Coin Selection algorithm might leak
```
Comment (Xekyo) [inline 2021-05-05-#17526.md:75]: Maybe useful for the notes: https://bitcoin.stackexchange.com/a/32445/5406

File: 2021-05-05-#17526.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+   information about the wallet's UTXO pool? Why do we
+[shuffle](https://github.com/bitcoin/bitcoin/blob/2b45cf0bcdb3d2c1de46899e30885c953b57b475/src/wallet/wallet.cpp#L2503)
+`vCoins` before creating `OutputGroup`s?
+
+8. Bonus: We've listed some qualitative (e.g. presence of a change output) and quantitative (e.g. number of inputs used) ways to
+compare Coin Selection solutions. Instead of returning as soon as `SelectCoinsMinConf()` finds a
+solution, should we try multiple and then pick one? How might we design a metric
+to decide which one to use?
```
Comment (Xekyo) [inline 2021-05-05-#17526.md:83]: Uuuh, nice already leading up to the future work in other PRs. :)

**Comment by glozow** [comment]
@jonatack OOPs, yes, somehow I didn't put it in _posts/ :scream:
Will fix asap

File: 2021-05-05-#17526.md (hunk 2 above)
Comment (glozow) [reply 2021-05-05-#17526.md:62]: Different codebase prolly

File: _posts/2021-05-05-#17526.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+* [PR #17526](https://github.com/bitcoin/bitcoin/pull/17526) implements Single Random Draw as
+  an additional fallback Coin Selection strategy. Single Random Draw is fairly straightforward: it
+randomly picks `OutputGroup`s from a pool of eligible UTXOs until the total amount is sufficient to
+cover the payment and fees. Any extra funds are put in a change output.
+
+* This means Coin Selection will have three different solvers: Branch and Bound (BnB), Knapsack, and
+  Single Random Draw (SRD). Note that some randomness is used in Coin Selection, so we won't always
+come up with the same solution.
+
+* The overall strategy within `SelectCoins()` (including PR #17331, on which PR #17526 is built)
```
Comment (jonatack) [inline _posts/2021-05-05-#17526.md:34]:
```suggestion
* The overall strategy within `SelectCoins()` (including PR #17331, on which PR #17526 is built):
```

File: _posts/2021-05-05-#17526.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	  `CoinEligibilityFilter`. There is a clear preference for BnB and we will only try Knapsack
+and SRD if that fails. However, we'll try both Knapsack and SRD together, picking the solution with
+lower fees, breaking ties by number of UTXOs used.
+
+
+## Questions
+
+1. Can you give a high-level description of the Coin Selection strategy proposed in this PR?
+
+2. Within `SelectCoinsMinConf()`, if we have a Knapsack and SRD solution, how do we decide which one
```
Comment (jonatack) [inline _posts/2021-05-05-#17526.md:56]:
```suggestion
2. Within `SelectCoinsMinConf()`, if we have both a Knapsack and an SRD solution, how do we decide which one
```

File: _posts/2021-05-05-#17526.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+offers an overview of tradeoffs between different Coin Selection strategies. We
+covered Coin Selection in a previous review club,
+[#17331](https://bitcoincore.reviews/17331).
+
+* [PR #17526](https://github.com/bitcoin/bitcoin/pull/17526) implements Single Random Draw as
+  an additional fallback Coin Selection strategy. Single Random Draw is fairly straightforward: it
+randomly picks `OutputGroup`s from a pool of eligible UTXOs until the total amount is sufficient to
+cover the payment and fees. Any extra funds are put in a change output.
+
+* This means Coin Selection will have three different solvers: Branch and Bound (BnB), Knapsack, and
```
Comment (jonatack) [inline _posts/2021-05-05-#17526.md:30]:
```suggestion
* This means that with this PR, our coin selection would have three different solvers: Branch and Bound (BnB), Knapsack, and
```

**Review by jonatack — COMMENTED** [review]
This looks fun! Great work. Maybe write "coin selection" instead of "Coin Selection" like we do in Optech (it's not in the style guide, but from grepping the optech codebase). A few suggestions to pick and choose from.

File: 2021-05-05-#17526.md (hunk 3 above)
Comment (glozow) [reply 2021-05-05-#17526.md]: Thanks! Added in the notes

File: _posts/2021-05-05-#17526.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+* **Coin Selection** refers to the process of selecting UTXOs (or "coins") from a wallet's available
+  UTXO pool to fund a transaction. Generally, the goal is to pick coins that will minimize
+fees for the user (now and in the long run), help the transaction reliably
+confirm in a timely manner, and not leak information about the wallet. [This
+Stack Exchange
+post](https://bitcoin.stackexchange.com/questions/32145/what-are-the-trade-offs-between-the-different-algorithms-for-deciding-which-utxo/32445#32445)
+offers an overview of tradeoffs between different Coin Selection strategies. We
+covered Coin Selection in a previous review club,
+[#17331](https://bitcoincore.reviews/17331).
```
[the remark this answers is not in the document]
Comment (glozow) [reply _posts/2021-05-05-#17526.md]: Ah! I will remember to do this from now on :D

File: _posts/2021-05-05-#17526.md (hunk 6 above)
Comment (glozow) [reply _posts/2021-05-05-#17526.md]: Done

**Comment by glozow** [comment]
Thanks @jonatack! Took your suggestions and changed to lowercase

**Comment by jnewbery** [comment]
Looks good. Thanks @glozow!

**Pull request merged**